18. Function Recursion

1. Calculating factorial

1.2 Factorial and loops

In [1]:
n = int(input("Enter a number : "))
fact = 1

for i in range(1,n+1) :
    fact = fact * i
    
print("{}! = {}".format(n,fact))

Enter a number : 5
5! = 120


1.3 Factorial and recursion

* The recursion refers to a function that calls itself from withing its     own body.
* Recursion is a very useful problem-solving technique that allows us to     solve complex problems in a intuitive and simple way, which be difficult   to solve using procedural techniques.

### The flow of the factorial recursion is as follows

factorial(5) 
    return 5 * factorial(4) = 120
        return 4 * factorial(3) = 24
            return 3 * factorial(2) = 6
                return 2 * factorial(1) = 2
                
* If factorial(1) returns 1, then factorial(2), which called factorial(1),follows the flow of '2 * factorial(1) = 2*1'.
* The final result obtained through repetitive returns is 120

In [10]:
def factorial(n) :
    if n <= 1:
        return 1
    else:
        return n * factorial(n-1)

    
n = 5
print('{}! = {}'.format(n,factorial(n)))


5! = 120


A termination condition is essential in a recursive function.

* The important point in recursive functions is when the program             terminates.
* According to the definition, if n is less than or equal to 1, we don't     need to define factorial for values less than 1.
* Therefore, if the 'n' in the if statement becomes 1 or less, we can       simply return 1.

### Recursive calls and divide-and-conquer
* Recursion is a technique where a function calls itself, and it is common   to break down a problem of size n into smaller subproblems of size less   than n and call them.
* This solving approach is closely related to the divide-and-conquer         algorithm.
* The divide-and-conquer algorithm can be considered as multiple-branch     recursion, since it only invloves a single recursive call.
* If the problem size is 'n', dividing the problem into n/2 and             recursively calling each will result in a typical divide-and-conquer       algorithm.
* There are also ways that divide the problem into more than two smaller     problems and apply recursive calls, All of these ways are condidered       divide-and-conquer algorithms.
* In divide-and-conquer problems, the sum of all the split problems should     not be larger than the original problem.

### 1.4 Pros and Cons of Recursive Calls and Loops

* Recursive Function
Pros : Relatively consise code
Cons : Requires more memory Relatively slower execution speed
* Loop
Pros : Relatively faster execution speed
Cons : Relatively complex code

### 2. Calculating Fibonacci sequence

#### 2.1 What is the Fibonacci sequence ?

* The first term of the Fibonacci sequence, F0, is 0, F1 is 1, and Fn is     defined as follows. Fn = Fn-1 + Fn-2
* According to this definition, the Fibonacci sequence is as follows.

### 2.2 Fibonacci sequence and recursion 

In [15]:
def fibonacci(n) :
    if n <= 1 :
        return n
    else:
        return fibonacci(n-1) + fibonacci(n-2)
    
nterms = int(input("How many Fibonacci number do you want?"))

if nterms <= 0 :
    print("Error : Enter positive number.")
else:
    print("Fibonacci Sequence : ",end="")
    for i in range(nterms) :
        print(fibonacci(i),end=" ")

How many Fibonacci number do you want?10
Fibonacci Sequence : 0 1 1 2 3 5 8 13 21 34 

* Let's examine the process of executing the fib(5) fibonacci function.
* fib(5) calls fib(4) and fib(3), and fib(4 recursivly calls fib(3) and     fib(2), resulting in a very large execution tree structure.

In [16]:
def fibonacci(n) :
    if n == 0 :
        return 0
    elif n == 1 :
        return 1 
    else :
        return fibonacci(n-1) + fibonacci(n-2)

print(fibonacci(10))

55


#### 2.3 Efficiency of Fibonacci Sequence Funtion Implemented with Recursive calls

* The biggest issue with recursive functions is their inefficiency because   they unnecessarily repeat calculations even for values that have already   been computed.
* Let's take the process of finding the 30th Fibonacci number as an         example.

In [20]:
from timeit import *

def fibonacci_1(n) :
    if n== 0 :
        return 0
    elif n == 1 :
        return 1
    else:
        return fibonacci_1(n-1) + fibonacci_1(n-2)
    
t3 = Timer("fibonacci_1(30)","from __main__ import fibonacci_1")
print("fibonacci_1(30) * 20 times : ",t3.timeit(number = 20),"seconds")

fibonacci_1(30) * 20 times :  2.895427292000022 seconds


#### 2.4 Fibonacci Sequence Function and Memoization

Memoization is an optimization technique that involves strong previously calculated values to avoid redundant calculations when they are repeated and their results remain unchanged.

In [12]:
from timeit import * 

dic = {0 : 0, 1 : 1}
def fibonacci_2(n) :
    if n in dic : 
        return dic[n]
    
    dic[n] = fibonacci_2(n-1) + fibonacci_2(n-2)
    return dic[n]

t2 = Timer("fibonacci_2(30)","from __main__ import fibonacci_2")
print("fibonacci_2(30) * 20 times",t2.timeit(number = 20), "seconds")

fibonacci_2(30) * 20 times 4.6208000185288256e-05 seconds


#### 18.3 Paper Coding

In [15]:
#Q1
n = int(input("Enter the number : "))

def sum_num(n) :
    if n == 1:
        return 1
    else :
        return n + sum_num(n-1)
    
print(sum_num(n))

Enter the number : 10
55


In [25]:
#Q2
from math import pow

x = int(input("Enter the values x : "))
n = int(input("Enter the values n : "))

def exponential (x,n) :
    if n == 1 :
        return pow(x,1)
    else :
        return x * exponential(x,n-1)
    
print(int(exponential(x,n)))

Enter the values x : 2
Enter the values n : 10
1024


#### 18.4 Let's code

In [31]:
def hanoi (n,from_,to_,via_) :
    if n== 1 : 
        print("Move disk {} from {} to {}".format(n,from_,to_))
    else :
        hanoi(n-1,from_,via_,to_)
        print("Move disk {} from {} to {}".format(n,from_,to_))
        hanoi(n-1,via_,to_,from_)
        
hanoi(3,'A','C','B')

Move disk 1 from A to C
Move disk 2 from A to B
Move disk 1 from C to B
Move disk 3 from A to C
Move disk 1 from B to A
Move disk 2 from B to C
Move disk 1 from A to C


#### 18.5 Pair Programming

In [54]:

def factorial(k) :
    if k <= 1 :
        return 1
    else :
        return k * factorial(k-1)
    
def euler(n) :
    if n == 1:
        return 1 
    else :
        return 1/factorial(n-1) + euler(n-1)
    
print(euler(20))

2.7182818284590455


In [55]:
#Q2
def gcd(a,b) :
    if b == 0 :
        return a
    else : 
        return (gcd(b,a%b))
    
a = int(input("Enter the number : "))
b = int(input("Enter the number : "))


print(gcd(a,b))

Enter the number : 48
Enter the number : 60
12
